Cleaning and exploring dataset on used Ebay Car Sales (classified section in German Ebay Website)
Data Quest Guided Project
7/20/2021
Iko Tumangan

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

VehicleType, gearbox, model, fuelType, notRepairedDamage are our columns with null values. We will have to handle these in some way

In [5]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.rename({'yearOfRegistration':'registration_year',
              'monthOfRegistration':'registration_month',
              'notRepairedDamage':'unrepaired_damage',
              'dateCreated':'ad_created'}, axis=1, inplace=True)

In [8]:
new_columns = []
for a in autos.columns:
    import re
    if a not in ["ad_created","unrepaired_damage","registration_month","registration_year"]:
        a = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', a)
        a = re.sub('([a-z0-9])([A-Z])', r'\1_\2', a).lower()
    new_columns.append(a)
autos.columns = new_columns

In [9]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

Made columns labels more clear, and standardized column label converting Camel Case to Snake Case

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

Seller, offer_type, abtest, gearbox, and unrepaired damage only have to two unique data points

In [11]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                  int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
kilometer              int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

In [12]:
autos.rename({'kilometer':'odometer_km'}, axis=1, inplace=True)

In [13]:
autos['price'].unique().shape

(5597,)

In [14]:
autos['price'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [15]:
autos['price'].value_counts()

0         10778
500        5670
1500       5394
1000       4649
1200       4594
          ...  
23456         1
171000        1
21830         1
13485         1
8188          1
Name: price, Length: 5597, dtype: int64

In [16]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

In [17]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

There are several anomalies with cars worth several million dollars. The most expensive cars in the world (Bugattis) are worth around $4 million. 

I am removing all cars valued above this mark. 

It's strange there are so many cars sold at the lower ranges, as low as $1 but it is possible to sell a used car for a dollar. 

There are also several cars worth $0, which is strange and can skew our actual analysis so I will be removing 

cars worth $0 dollars as well.

In [18]:
autos = autos[autos['price'].between(1, 4000000)]

In [19]:
autos['price'].describe()

count    3.606980e+05
mean     6.068401e+03
std      1.818741e+04
min      1.000000e+00
25%      1.250000e+03
50%      3.000000e+03
75%      7.490000e+03
max      3.895000e+06
Name: price, dtype: float64

In [20]:
autos['price'].value_counts().sort_index(ascending=False).head()

3895000    1
3890000    1
2995000    1
2795000    1
1600000    2
Name: price, dtype: int64

In [21]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head(20)

150000    233087
125000     37372
100000     15482
90000      12350
80000      10905
70000       9673
60000       8593
50000       7534
40000       6320
30000       5937
20000       5535
10000       1873
5000        6037
Name: odometer_km, dtype: int64

odomter_km values look fine. There doesn't seem to be any major outliers
The majority of used cars sold have 150000 km or more.
There are still some unrealisticly expensive cars sold on Ebay, but it is plausible that there are cars worth 3 or 4 million dollars.

In [22]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

In [23]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025545
2016-03-06    0.014480
2016-03-07    0.035653
2016-03-08    0.033466
2016-03-09    0.034112
2016-03-10    0.032648
2016-03-11    0.032775
2016-03-12    0.036244
2016-03-13    0.015781
2016-03-14    0.036327
2016-03-15    0.033424
2016-03-16    0.030200
2016-03-17    0.031655
2016-03-18    0.013119
2016-03-19    0.035268
2016-03-20    0.036399
2016-03-21    0.035695
2016-03-22    0.032493
2016-03-23    0.032002
2016-03-24    0.029914
2016-03-25    0.032798
2016-03-26    0.031969
2016-03-27    0.030230
2016-03-28    0.035060
2016-03-29    0.034123
2016-03-30    0.033529
2016-03-31    0.031874
2016-04-01    0.034150
2016-04-02    0.035096
2016-04-03    0.038811
2016-04-04    0.037630
2016-04-05    0.012786
2016-04-06    0.003127
2016-04-07    0.001619
Name: date_crawled, dtype: float64

date_crawled covers roughly a month, and seems evenly distributed throughout the month

In [24]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.038999
2016-04-04    0.037738
2016-04-05    0.011619
2016-04-06    0.003119
2016-04-07    0.001555
Name: ad_created, Length: 114, dtype: float64

ad_created data spans over a little more than two years. There are smaller percentages of data at each ad_created date point

so less data entries by ad_created date compared to 'date_crawled'

In [25]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005201
2016-03-08    0.007940
2016-03-09    0.009823
2016-03-10    0.011458
2016-03-11    0.012953
2016-03-12    0.023241
2016-03-13    0.008409
2016-03-14    0.012174
2016-03-15    0.016327
2016-03-16    0.016415
2016-03-17    0.028703
2016-03-18    0.006892
2016-03-19    0.016332
2016-03-20    0.019889
2016-03-21    0.020031
2016-03-22    0.020510
2016-03-23    0.018018
2016-03-24    0.019163
2016-03-25    0.018997
2016-03-26    0.015955
2016-03-27    0.016720
2016-03-28    0.022190
2016-03-29    0.023285
2016-03-30    0.023721
2016-03-31    0.024239
2016-04-01    0.023898
2016-04-02    0.024971
2016-04-03    0.025320
2016-04-04    0.025531
2016-04-05    0.126968
2016-04-06    0.218942
2016-04-07    0.130422
Name: last_seen, dtype: float64

Last seen also spans for about a month like 'date_crawled.' There are far more data points in just the last 3 days 

In [26]:
autos['registration_year'].describe()

count    360698.000000
mean       2004.451877
std          82.110038
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

Really strange to see min year to be 1000 and max year to 9999. No cars yet in year 1000 and year 9999 is way out in the future

In [27]:
autos['registration_year'].value_counts().sort_index(ascending=True).head(20)

1000    32
1001     1
1039     1
1111     2
1234     4
1255     1
1300     2
1400     1
1500     3
1600     2
1602     1
1800     5
1910    58
1911     1
1923     3
1925     1
1927     2
1928     2
1929     8
1930     4
Name: registration_year, dtype: int64

In [28]:
autos['registration_year'].value_counts().sort_index(ascending=True).tail(20)

5300     1
5555     2
5600     1
5900     1
5911     2
6000     6
6200     1
6500     1
7000     4
7100     1
7500     1
7777     1
7800     1
8000     2
8200     1
8500     1
8888     2
9000     4
9450     1
9999    19
Name: registration_year, dtype: int64

Removing outliers and only keeping rows with registration year between 1900 and 2016

In [29]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000167
1911    0.000003
1923    0.000009
1925    0.000003
1927    0.000006
          ...   
2012    0.026993
2013    0.017657
2014    0.013755
2015    0.008430
2016    0.026612
Name: registration_year, Length: 94, dtype: float64

In [30]:
autos['registration_year'].describe()

count    346720.000000
mean       2002.895737
std           7.247339
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        2016.000000
Name: registration_year, dtype: float64

In [31]:
autos['registration_year'].value_counts().sort_index()

1910      58
1911       1
1923       3
1925       1
1927       2
        ... 
2012    9359
2013    6122
2014    4769
2015    2923
2016    9227
Name: registration_year, Length: 94, dtype: int64

Most cars sold were cars registered in 1999 and 2000

In [32]:
autos['brand'].value_counts()

volkswagen        73392
bmw               38094
opel              36890
mercedes_benz     33574
audi              31044
ford              23893
renault           16472
peugeot           10453
fiat               8906
seat               6469
skoda              5438
mazda              5333
smart              4968
citroen            4836
nissan             4714
toyota             4483
hyundai            3457
sonstige_autos     3310
mini               3253
volvo              3171
mitsubishi         2855
honda              2611
kia                2397
suzuki             2207
alfa_romeo         2187
porsche            2170
chevrolet          1742
chrysler           1339
dacia               865
jeep                761
land_rover          751
daihatsu            749
subaru              734
jaguar              602
saab                508
daewoo              505
trabant             488
lancia              451
rover               441
lada                207
Name: brand, dtype: int64

In [33]:
autos['brand'].value_counts(normalize=True, ascending=False)

volkswagen        0.211675
bmw               0.109870
opel              0.106397
mercedes_benz     0.096833
audi              0.089536
ford              0.068912
renault           0.047508
peugeot           0.030148
fiat              0.025686
seat              0.018658
skoda             0.015684
mazda             0.015381
smart             0.014329
citroen           0.013948
nissan            0.013596
toyota            0.012930
hyundai           0.009971
sonstige_autos    0.009547
mini              0.009382
volvo             0.009146
mitsubishi        0.008234
honda             0.007531
kia               0.006913
suzuki            0.006365
alfa_romeo        0.006308
porsche           0.006259
chevrolet         0.005024
chrysler          0.003862
dacia             0.002495
jeep              0.002195
land_rover        0.002166
daihatsu          0.002160
subaru            0.002117
jaguar            0.001736
saab              0.001465
daewoo            0.001457
trabant           0.001407
l

The above shows the breakdown of cars sold by car brand. For this analysis we are going to take the top 24 car brands and perform our aggregation and mean calculation.

In [34]:
brand_select = autos['brand'].value_counts().nlargest(24).index
print(brand_select)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'smart', 'citroen',
       'nissan', 'toyota', 'hyundai', 'sonstige_autos', 'mini', 'volvo',
       'mitsubishi', 'honda', 'kia', 'suzuki'],
      dtype='object')


In [35]:
mean_price_dict = {}

for c in brand_select:
    brand = autos[autos['brand'] == c]
    mean_price = brand['price'].mean()
    mean_price_dict[c] = int(mean_price)
    
mean_price_dict

{'volkswagen': 5452,
 'bmw': 8573,
 'opel': 3028,
 'mercedes_benz': 8609,
 'audi': 9166,
 'ford': 3779,
 'renault': 2437,
 'peugeot': 3267,
 'fiat': 2892,
 'seat': 4541,
 'skoda': 6530,
 'mazda': 4076,
 'smart': 3632,
 'citroen': 3734,
 'nissan': 4708,
 'toyota': 5339,
 'hyundai': 5567,
 'sonstige_autos': 22155,
 'mini': 10080,
 'volvo': 5238,
 'mitsubishi': 3407,
 'honda': 4005,
 'kia': 5855,
 'suzuki': 4495}

In [36]:
mean_mileage_dict = {}

for d in brand_select:
    brand = autos[autos['brand'] == d]
    mean_mileage = brand['odometer_km'].mean()
    mean_mileage_dict[d] = int(mean_mileage)
    
mean_mileage_dict

{'volkswagen': 128387,
 'bmw': 132791,
 'opel': 128721,
 'mercedes_benz': 130566,
 'audi': 129440,
 'ford': 123663,
 'renault': 127885,
 'peugeot': 124500,
 'fiat': 116619,
 'seat': 120608,
 'skoda': 113593,
 'mazda': 125754,
 'smart': 99505,
 'citroen': 120214,
 'nissan': 119446,
 'toyota': 117271,
 'hyundai': 104080,
 'sonstige_autos': 87179,
 'mini': 93273,
 'volvo': 138131,
 'mitsubishi': 126959,
 'honda': 125748,
 'kia': 109144,
 'suzuki': 106726}

In [37]:
price_series = pd.Series(mean_price_dict)
df = pd.DataFrame(price_series, columns=['mean_price'])

mileage_series = pd.Series(mean_mileage_dict)
df['mean_mileage'] = mileage_series

counts_series = autos['brand'].value_counts()
df['count_sold'] = counts_series

df

mean_price  mean_mileage  count_sold
volkswagen            5452        128387       73392
bmw                   8573        132791       38094
opel                  3028        128721       36890
mercedes_benz         8609        130566       33574
audi                  9166        129440       31044
ford                  3779        123663       23893
renault               2437        127885       16472
peugeot               3267        124500       10453
fiat                  2892        116619        8906
seat                  4541        120608        6469
skoda                 6530        113593        5438
mazda                 4076        125754        5333
smart                 3632         99505        4968
citroen               3734        120214        4836
nissan                4708        119446        4714
toyota                5339        117271        4483
hyundai               5567        104080        3457
sonstige_autos       22155         87179        3310
mini                 10080         93273        3253
volvo                 5238        138131        3171
mitsubishi            3407        126959        2855
honda                 4005        125748        2611
kia                   5855        109144        2397
suzuki                4495        106726        2207

The top 3 selling cars brands are volkswagen, bmw, and opel (volkswagen by far). The mean price of volkswagen cars sold are $5388 and 128883km respectively. Bmw means price is significantly higher with a slight higher total mileage driven, opel relatively chepaer. The mean_mileage doe not vary as much as price, and some car brands are relatively more expensive despite having higher or similar miles driven. Just looking at the top 6 car brands sold, we see that bmw, mercedes, and audi have a much higher price compared to opel and ford. Volkswagen the highest selling car is priced in the middle. 

Analysis next steps:
1) Find the most common brand/model combinations
2) Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
3) How much cheaper are cars with damage than their non-damaged counterparts?

In [38]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [39]:
odo_groups = autos['odometer_km'].value_counts()
print(odo_groups)

150000    223023
125000     36010
100000     15039
90000      11988
80000      10612
70000       9440
60000       8434
50000       7362
40000       6214
30000       5803
5000        5632
20000       5375
10000       1788
Name: odometer_km, dtype: int64


In [40]:
odo_index = odo_groups[odo_groups > 0].index
odo_index

Int64Index([150000, 125000, 100000,  90000,  80000,  70000,  60000,  50000,
             40000,  30000,   5000,  20000,  10000],
           dtype='int64')

In [41]:
odo_dict = {} 

for o in odo_index:
    odo = autos[autos['odometer_km'] == o]
    mean_price = odo['price'].mean()
    odo_dict[o] = int(mean_price)

#sorting dictionary by key
sorted_odo_dict = {k: v for k, v in sorted(odo_dict.items(), key=lambda item: item[0])}
sorted_odo_dict

{5000: 13899,
 10000: 23976,
 20000: 17751,
 30000: 16212,
 40000: 15270,
 50000: 13837,
 60000: 12138,
 70000: 10976,
 80000: 9690,
 90000: 8732,
 100000: 8169,
 125000: 6358,
 150000: 3868}

In [42]:
odo_series = pd.Series(sorted_odo_dict)
df = pd.DataFrame(odo_series, columns=['mean_price'])
df

mean_price
5000         13899
10000        23976
20000        17751
30000        16212
40000        15270
50000        13837
60000        12138
70000        10976
80000         9690
90000         8732
100000        8169
125000        6358
150000        3868

Clear to see price generally decreasing as mileage increases. We see prices are quite small for the low mileage of 5,000. We might have to dig deeper to find out why this is the case. There could also be relatively low data points for this mileage range.

In [44]:
brand_list = []

for x in autos['brand']:
    if x not in brand_list:
        brand_list.append(x)
        
brand_list

['volkswagen',
 'audi',
 'jeep',
 'skoda',
 'bmw',
 'peugeot',
 'ford',
 'mazda',
 'nissan',
 'renault',
 'mercedes_benz',
 'seat',
 'honda',
 'fiat',
 'opel',
 'mini',
 'smart',
 'hyundai',
 'sonstige_autos',
 'alfa_romeo',
 'subaru',
 'volvo',
 'mitsubishi',
 'kia',
 'suzuki',
 'lancia',
 'porsche',
 'citroen',
 'toyota',
 'chevrolet',
 'dacia',
 'daihatsu',
 'trabant',
 'chrysler',
 'jaguar',
 'daewoo',
 'rover',
 'saab',
 'land_rover',
 'lada']

In [45]:
brand_model_dict = {}

for a in brand_list:
    selected_rows = autos[autos['brand'] == a]
    model_select = selected_rows['model'].value_counts(dropna=False)
    model_select = str(model_select)
    model_select_split = model_select.split()
    model_name = model_select_split[0]
    brand_model_dict[a] = model_name

In [46]:
brand_model_series = pd.Series(brand_model_dict)
brand_model_df = pd.DataFrame(brand_model_series, columns=['model_type'])
brand_model_df

model_type
volkswagen            golf
audi                    a4
jeep                 grand
skoda              octavia
bmw                    3er
peugeot            2_reihe
ford                 focus
mazda              6_reihe
nissan               micra
renault             twingo
mercedes_benz     c_klasse
seat                 ibiza
honda                civic
fiat                 punto
opel                 corsa
mini                cooper
smart               fortwo
hyundai            i_reihe
sonstige_autos         NaN
alfa_romeo             156
subaru             impreza
volvo                  v40
mitsubishi            colt
kia                 andere
suzuki              andere
lancia             ypsilon
porsche                911
citroen             andere
toyota               yaris
chevrolet           andere
dacia                logan
daihatsu             cuore
trabant                601
chrysler            andere
jaguar              andere
daewoo               matiz
rover               andere
saab                andere
land_rover      freelander
lada                  niva

In [47]:
brand_model_dict_volumes = {}

for a in brand_list:
    selected_rows = autos[autos["brand"] == a]
    model_select = selected_rows["model"].value_counts(dropna=False)
    model_select = str(model_select)
    model_select_split = model_select.split()
    model_volumes = model_select_split[1]
    model_volumes = int(model_volumes)
    brand_model_dict_volumes[a] = model_volumes
    
model_series_vol = pd.Series(brand_model_dict_volumes)
# print(model_series_vol)
brand_model_vol_df = pd.DataFrame(model_series_vol, columns=['model_volumes'])
# brand_model_vol_df
brand_model_df["model_volumes"] = model_series_vol
brand_model_df_vol_sorted = brand_model_df.sort_values("model_volumes", ascending=False)
# brand_model_df_vol_sorted
brand_model_df_vol_sorted

model_type  model_volumes
volkswagen            golf          27548
bmw                    3er          19470
opel                 corsa          11598
audi                    a4           9800
mercedes_benz     c_klasse           8444
ford                 focus           5597
peugeot            2_reihe           4646
renault             twingo           4524
smart               fortwo           4156
sonstige_autos         NaN           3310
fiat                 punto           3043
seat                 ibiza           2505
skoda              octavia           2136
mini                cooper           2061
citroen             andere           1742
nissan               micra           1634
mazda              6_reihe           1247
honda                civic           1231
hyundai            i_reihe           1228
suzuki              andere           1009
toyota               yaris            991
chevrolet           andere            984
porsche                911            935
mitsubishi            colt            809
volvo                  v40            723
alfa_romeo             156            590
kia                 andere            470
chrysler            andere            369
rover               andere            367
saab                andere            356
trabant                601            354
daihatsu             cuore            353
dacia                logan            313
jaguar              andere            297
jeep                 grand            273
subaru             impreza            218
land_rover      freelander            214
lancia             ypsilon            191
daewoo               matiz            148
lada                  niva            124

We can see which model types above are selling the most

In [48]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [49]:
autos['unrepaired_damage'].value_counts()

nein    252813
ja       33191
Name: unrepaired_damage, dtype: int64

In [50]:
brand_damage_dict = {}

for a in brand_list:
    selected_rows = autos.loc[(autos['brand'] == a) & ((autos['unrepaired_damage'] == 'ja'))]
    damage_mean_price = selected_rows['price'].mean()
    brand_damage_dict[a] = int(damage_mean_price)
    
brand_damage_series = pd.Series(brand_damage_dict)
brand_damage_df = pd.DataFrame(brand_damage_series, columns=['mean_price_damaged'])
brand_damage_df

mean_price_damaged
volkswagen                    2538
audi                          3419
jeep                          4230
skoda                         3279
bmw                           3660
peugeot                       1453
ford                          1696
mazda                         1601
nissan                        1719
renault                       1016
mercedes_benz                 3749
seat                          1915
honda                         1742
fiat                          1304
opel                          1251
mini                          5237
smart                         1270
hyundai                       2252
sonstige_autos                5060
alfa_romeo                    1557
subaru                        2033
volvo                         1724
mitsubishi                    1383
kia                           2107
suzuki                        1868
lancia                        1337
porsche                      19159
citroen                       1756
toyota                        3192
chevrolet                     3351
dacia                         3382
daihatsu                       886
trabant                        758
chrysler                      1566
jaguar                        3861
daewoo                         621
rover                          779
saab                          1072
land_rover                    5550
lada                          1732

In [51]:
brand_damage_repaired = {}

for a in brand_list:
    selected_rows = autos.loc[(autos['brand'] == a) & (autos['unrepaired_damage'] == 'nein')]
    repaired_mean_price = selected_rows['price'].mean()
    brand_damage_repaired[a] = int(repaired_mean_price)

repaired_series = pd.Series(brand_damage_repaired)
repaired_df = pd.DataFrame(repaired_series, columns = ['mean_price_not_damaged'])

brand_damage_df['mean_price_not_damaged'] = repaired_series
brand_damage_df_sorted = brand_damage_df.sort_values(by='mean_price_not_damaged', ascending=False)
brand_damage_df_sorted

mean_price_damaged  mean_price_not_damaged
porsche                      19159                   50899
sonstige_autos                5060                   29798
land_rover                    5550                   19383
jaguar                        3861                   17300
jeep                          4230                   14166
mini                          5237                   10611
audi                          3419                   10598
mercedes_benz                 3749                    9706
bmw                           3660                    9657
chevrolet                     3351                    7694
skoda                         3279                    7187
kia                           2107                    6744
volkswagen                    2538                    6429
dacia                         3382                    6326
volvo                         1724                    6308
hyundai                       2252                    6307
toyota                        3192                    5766
nissan                        1719                    5683
subaru                        2033                    5605
seat                          1915                    5313
suzuki                        1868                    5300
alfa_romeo                    1557                    5126
chrysler                      1566                    5053
mazda                         1601                    4907
honda                         1742                    4788
saab                          1072                    4672
ford                          1696                    4647
citroen                       1756                    4334
mitsubishi                    1383                    4186
lancia                        1337                    4156
smart                         1270                    4000
peugeot                       1453                    3851
lada                          1732                    3811
opel                          1251                    3639
fiat                          1304                    3551
renault                       1016                    3037
trabant                        758                    2924
daihatsu                       886                    2153
rover                          779                    2126
daewoo                         621                    1201

In [52]:
brand_damage_df['pct_diff'] = ((brand_damage_df['mean_price_damaged'] / brand_damage_df['mean_price_not_damaged']) * 100).astype(int)
brand_damage_df_final_sorted = brand_damage_df.sort_values(by='pct_diff', ascending = False)
brand_damage_df_final_sorted

mean_price_damaged  mean_price_not_damaged  pct_diff
toyota                        3192                    5766        55
dacia                         3382                    6326        53
daewoo                         621                    1201        51
mini                          5237                   10611        49
lada                          1732                    3811        45
skoda                         3279                    7187        45
chevrolet                     3351                    7694        43
daihatsu                       886                    2153        41
citroen                       1756                    4334        40
volkswagen                    2538                    6429        39
mercedes_benz                 3749                    9706        38
bmw                           3660                    9657        37
peugeot                       1453                    3851        37
porsche                      19159                   50899        37
seat                          1915                    5313        36
rover                          779                    2126        36
ford                          1696                    4647        36
subaru                        2033                    5605        36
honda                         1742                    4788        36
fiat                          1304                    3551        36
suzuki                        1868                    5300        35
hyundai                       2252                    6307        35
opel                          1251                    3639        34
renault                       1016                    3037        33
mitsubishi                    1383                    4186        33
mazda                         1601                    4907        32
lancia                        1337                    4156        32
audi                          3419                   10598        32
smart                         1270                    4000        31
kia                           2107                    6744        31
nissan                        1719                    5683        30
alfa_romeo                    1557                    5126        30
chrysler                      1566                    5053        30
jeep                          4230                   14166        29
land_rover                    5550                   19383        28
volvo                         1724                    6308        27
trabant                        758                    2924        25
jaguar                        3861                   17300        22
saab                          1072                    4672        22
sonstige_autos                5060                   29798        16

toyota, dacia, daewoo, mini, lada, skoda are the models that have the largest pecentage of units sold damaged versus undamaged. Clear to see that the mean price of damaged cars are much lower than that that are repaired or undamaged. This is expected.	